File: lib/src/main/java/com/github/aakumykov/android_dynamic_shortcuts_manager/dynamic_shortcut_manager/DynamicShortcutManager.java
package com.github.aakumykov.android_dynamic_shortcuts_manager.dynamic_shortcut_manager;

import android.content.Context;

import androidx.core.content.pm.ShortcutInfoCompat;
import androidx.core.content.pm.ShortcutManagerCompat;

import com.github.aakumykov.android_dynamic_shortcuts_manager.shortcuts_parser.model.Shortcut;

import java.util.ArrayList;
import java.util.List;

public class DynamicShortcutManager {

    // TODO: ShortcutManagerCompat.getMaxShortcutCountPerActivity()

    private final Context context;

    public DynamicShortcutManager(Context context) {
        this.context = context;
    }

    public void createDynamicShortcuts(List<Shortcut> list) throws IllegalArgumentException {
        List<ShortcutInfoCompat> shortcutInfoCompatList = new ArrayList<>();
        for (Shortcut shortcut : list) {
            shortcutInfoCompatList.add(shortcut.toShortcutInfo(context));
        }
        ShortcutManagerCompat.addDynamicShortcuts(context, shortcutInfoCompatList);
    }

    public void removeDynamicShortcuts(List<Shortcut> list) {
        List<String> idList = new ArrayList<>();
        for (Shortcut shortcut : list) {
            idList.add(shortcut.shortcutId);
        }
        ShortcutManagerCompat.removeDynamicShortcuts(context, idList);
    }
}

File: lib/src/main/java/com/github/aakumykov/android_dynamic_shortcuts_manager/dynamic_shortcut_manager/DynamicShortcutManager.kt
package com.github.aakumykov.android_dynamic_shortcuts_manager.dynamic_shortcut_manager

import android.content.Context
import androidx.core.content.pm.ShortcutManagerCompat
import com.github.aakumykov.android_dynamic_shortcuts_manager.shortcuts_parser.model.ShortcutJava

class DynamicShortcutManager(private val context: Context) {

    /**
     * @param List of [ShortcutJava] objects. Must be <= 4.
     * @return Number of created shortcuts.
     */
    fun createDynamicShortcuts(list: List<ShortcutJava>): Result<Int> {
        return list.map { shortcutList ->
            shortcutList.toShortcutInfo(context)
        }
            .let { shortcutInfoList ->
                try {
                    ShortcutManagerCompat.addDynamicShortcuts(context, shortcutInfoList)
                    Result.success(shortcutInfoList.size)
                }
                catch (e: Exception) {
                    Result.failure(e)
                }
            }
    }

    fun removeDynamicShortcuts(list: List<ShortcutJava>) {
        list.map { it.shortcutId }.also { idList ->
            ShortcutManagerCompat.removeDynamicShortcuts(context, idList)
        }
    }
}